dali/dfu/dfurun.cpp

```diff
@@ -601,6 +601,19 @@ class CDFUengine: public CInterface, implements IDFUengine
             throw MakeStringException(DFSERR_CreateAccessDenied,"Create permission denied for physical file(s)");
     }
 
+    void checkPlaneFilePermissions(IFileDescriptor *fd,IUserDescriptor *user,bool write)
```

Comment:
I don't think this is the right approach, or adding getPlaneFilePermissions based on an IFileDescriptor to IDistributedFileDirectory.

The pre-existing getFDescPermissions in the long run should be retired too. e.g. Foreign files should also not rely on an IFileDescriptor just to check the permissions (should be able to check the permissons of a foreign::ip::paths without an IFileDescriptor). But that is beyond the scope of this PR.

We already have getDropZoneScopePermissions for when a plane + paths is known, or getDLFNPermissions for generalized CDfsLogicalFileName scope checking, they should be reused.

If the plane is know (supplied to the DFU workunit via source->setGroupName), then it should use it directly and the filepath (or filepaths if multi), using 1 of the pre-existing calls.

If not, then it can callback on a legacy to checkPhysicalFilePermissions.
NB: there should be comment that this is legacy only, and it should be a configurable block default on in bare-metal, off in cloud - in the long run we should get rid of checkPhysicalFilePermissions altogether. 

NB: in the multi case, we already don't support a mixture of plane, non-plane filenames, so no attempt should be made to support both here. Either a plane is supplied or not (which implies bare-metal and badly configured dropzones with restrictions off).

Reply:
Also, if you're having to parse multi filenames here, I suspect it's going to be almost identical code to that in checkDZScopeAccessAndSetSpraySourceDFUFileSpec.

If so , to avoid duplication, multiple paths and make it easier to maintain the code going forward/avoid different problems in different implementations, this code should be commoned up and used by both areas.

Reply:
@jakesmith I am using the IFileDescriptor to find out the planeName (GroupName). I do have to find out and parse multi filenames. I am also using the IFileDescriptor to find out the multi filenames using the getMultiFilename(). In dfuserver code, is there a better way to find out the planeName and multi filenames without using the IFileDescriptor?

Reply:
Are you suggesting that moving the code from the getPlaneFilePermissions() to the checkPlaneFilePermissions() and calling the getDropZoneScopePermissions() directly in the checkPlaneFilePermissions()?

Reply:
>@jakesmith I am using the IFileDescriptor to find out the planeName (GroupName).

the plane name is parsed (should be) in the IDFUfileSpec spec. as part of the DFU workunit, see:
```
        source->setGroupName(sourcePlaneReq);
```
in CFileSprayEx::onSprayFixed

Reply:
>In dfuserver code, is there a better way to find out the planeName and multi filenames without using the IFileDescriptor?

the service does without a IFileDescriptor, DFU has the list of files (if multiple), there is already code in the service that parses this and checks dropzone+each file's permissions.
DFU server and service should use common code

Reply:
>Are you suggesting that moving the code from the getPlaneFilePermissions() to the checkPlaneFilePermissions() and calling the getDropZoneScopePermissions() directly in the checkPlaneFilePermissions()?

Not really.

I am suggsting that the plane is available (except in weird cases with restrictions off, in which case it should behave exactly as it already does now), but if plane then it should do the same as the service code does (and use common code).
It should introduce any functions or use any functions based on IFileDescriptors.

Reply:
> the service does without a IFileDescriptor, DFU has the list of files (if multiple), there is already code in the service that parses this and checks dropzone+each file's permissions. DFU server and service should use common code

In ESP FileSpray service, the list of files (if multiple) is in a comma separated string. I did research and found that I may be able to build the list of files using getRawDirectory() and getRawFileMask(). Am I in a correct direction?

Reply:
ok, yuck, the dfu workunit interface is pretty horrible in this respect, it's very geared towards carrying meta data suitable for a IFileDescriptor, making it difficult to get the simple underlying filename out in the simple cases and/or non logical file case.

Since it also always deals with an IFileDescriptor, it is not worth trying to build up the source filenames in any other way.

However, we should add a comment block to checkPlaneFilePermissions explaining why and what it's doing with the IFileDescriptor in this context).

The impl. should expect a dropzone to be populated, if not, it implies it is due to it not matching a dropzone with useDropZoneRestriction=false.
And, in that case , it should check a configurable option, and call checkPhysicalFilePermissions

The current impl. of checkPlaneFilePermissions is checking for the multi case, and looping over every file, but that is not necessary, since the only supported form of wildcard requires all files to be in the same directory.
So it is only necessary to scope check the directory afaics.

Reply:
The list of files received by the FileSpray service is:
//127.0.0.1/home/lexis/runtime/var/lib/HPCCSystems/mydropzone/subfolder1/file1,//127.0.0.1/home/lexis/runtime/var/lib/HPCCSystems/mydropzone/subfolder1/file2.
The getRawDirectory() returns: /home/lexis/runtime/var/lib/HPCCSystems/mydropzone/subfolder1.
The getRawFileMask() returns: file1,file2.

Reply:
Looks like it is enough to only check the first file if multi.